### ISSUE TITLE
RUCSS - Safe mode to not process Inline CSS

### ISSUE THREAD

**Issue by mehedihasanziku** [wp-media/wp-rocket#4166]
**Is your feature request related to a problem? Please describe.**
There is no site probably exist in at least wordpress that they don't have inline css using `<style>` tag. so yes if we use any inline css that must be needed to update sometimes for sure. so now the problem is we need to `"Clear Used CSS"` everytime whenever we do some changes with inline css. before if we do `"Clear Cache"` it clears the html so inline css was updated, but now clear cache doesn't trigger `"Clear Used CSS"`, so inline css not updating here.

**Describe the solution you'd like**
I want an safe mode option to not proccess inline css.

**Describe alternatives you've considered**
If you don't want to add option more, then I will be happy if you guard this with an wordpress filter, it will be also okay for me. so i can debug easily.

**Additional context**
N/A

**Comment by GeekPress** [comment]
@Tabrisrp Is this possible to do?

**Comment by remyperona** [comment]
Clearing the used CSS after doing the change will also clear the cache, so the end result is the same. The new inline CSS will be processed immediately on the next page cache.

**Comment by mehedihasanziku** [comment]
Yes, it does.

but problem happens when some dynamic class changed every page load, like (tagdiv composer), so after a period when preload cache/clear(cache lifespan) happened html updated but rucss not(as it will clear after 30days), that time user will see broken style till not press "Clear Used CSS"

**Comment by vmanthos** [comment]
I noticed the same when the `id` of an element changes dynamically.

This was on a Flatsome website where ids like `banner-959469755` are used:
![](https://i.vgy.me/nTztk1.jpg)

The CSS for those are added inline:
```CSS
<style>
#banner-959469755 {
  padding-top: 100%;
  background-color: rgb(255, 255, 255);
}
#banner-959469755 .bg.bg-loaded {
  background-image: url(https://www.aja-nantes.com/wp-content/uploads/2021/07/photo-accueil-07-2021.jpg);
}
</style>
```
When caching isn't in place the ids change on each page load. With caching they change every time the cache is cleared.

I safelisted `banner-(.*)` and the rules were added to the used CSS. However, when the cache was cleared, and consequently the id changed in the HTML, the used CSS still pointed to the ID that was in place when it was created.

Clearing the used CSS makes this work, until the next time the cache is cleared.

So, the pattern is this:
1. Used CSS is created - it targets the correct dynamic id, e.g. `banner-A` => The banner works.
2. Cache is cleared - HTML now has `banner-B` for the same banner, and used CSS points to the initial `banner-A` . => The banner breaks.
3. Clear used CSS - HTML and used CSS are aligned =>The banner works.

**ticket:** https://secure.helpscout.net/conversation/1589092258/284541/

**Comment by vmanthos** [comment]
I had another case, again with Flatsome's dynamic `banner-` elements. 

The SaaS does find the inline CSS, so it will be there even if `banner-(.*)` is not safelisted. It will not add the updated `id` when tree-shaking occurs the next time the page is sent to the SaaS.

**Ticket:** https://secure.helpscout.net/conversation/1590220890/284964/

**Comment by vmanthos** [comment]
The same thing happens on a site using Unlimited Elements for Elementor (Premium).

That's adding inline CSS to style `<ul>`:
```CSS
#uc_simple_list_menu_elementor34678 ul {
    padding: 0;
    list-style: none;
    margin: 0;
}
```

This is how the list normally looks:
![](https://i.vgy.me/1zahwv.jpg)

The id changes every time the cache is cleared and the list's style breaks:
![](https://i.vgy.me/pmNCNc.jpg)


**Ticket:** https://secure.helpscout.net/conversation/1586450973/283902?folderId=2135277

**Comment by mehedihasanziku** [comment]
related #4265
 
**` Noticed In Thrive Builder`**

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1621682981/292341?folderId=2135277

This is about a **Flatsome** slider. They are using Flickity, and the background of the slider is added using an inline, dynamic `#banner-` id:
```HTML
<style>
#banner-1787931694 .bg.bg-loaded {
  background-image: url(https://faunasalud.cl/wp-content/uploads/2021/05/Carrusel-version-2-reptil-1.jpg);
}
#banner-1787931694 .bg {
  background-position: 50% 50%;
}
```
The first time the used CSS is created the slider is being displayed. Clearing the cache breaks the slider.

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1631546552/294820?folderId=2135277

In this case, it's the [Ronneby theme](https://themeforest.net/item/ronneby-highperformance-wordpress-theme/11776839) that uses JavaScript to add a `<style></style>` block containing ids that change on every (non-cached) page load:

![](https://i.vgy.me/mhCrTx.jpg)

**Comment by vmanthos** [comment]
Related ticket: https://secure.helpscout.net/conversation/1639266873/296590?folderId=2135277

This was a Flickity slider used in Flatsome. Again the pattern was like this:
```CSS
<style>
#banner-1787931694 .bg.bg-loaded {
  background-image: url(https://example.com/wp-content/uploads/2021/05/Carrusel-version-2-reptil-1.jpg);
}
#banner-1787931694 .bg {
  background-position: 50% 50%;
}
```

**Comment by DahmaniAdame** [comment]
Related - https://secure.helpscout.net/conversation/1661281740/301228/

**Comment by mehedihasanziku** [comment]
Here can be introduced another attribute like **Delay JS** `"nowprocket"`
> https://github.com/wp-media/wp-rocket/blob/860c576320af31ff84cb8c5d0bcf9d269ebe9fbe/inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php#L82

It will be good in some cases, like client will see effect immediately.

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1677955163/304285/

**Comment by vmanthos** [comment]
Related: https://secure.helpscout.net/conversation/1702907838/309065/

This is related to Flatsome's dynamic ids/classes.

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1701180267/308771/#thread-5017465027

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1708424786/310639?folderId=2135277

It's a Flatsome slider this time named: `banner-`+`random numbers`.

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1728256924/314490/

**Comment by vmanthos** [comment]
**Related ticket:** https://secure.helpscout.net/conversation/1726370873/314178?folderId=273766

In this case, there are dynamic ids added by the [Unlimited Elements for Elementor (Premium)](https://unlimited-elements.com/) plugin:
```HTML
<div id="uc_post_magazine_grid_elementor15161" class="post_magazine_grid ue-layout-cover uc-items-wrapper">
```

**Comment by vmanthos** [comment]
Related ticket: https://secure.helpscout.net/conversation/1790500857/326787?folderId=273766

This is another case of Flatsome's `#banner-(.*)` dynamic selector.

**Comment by NataliaDrause** [comment]
Related: https://secure.helpscout.net/conversation/1795001410/327658/

Flatsome's `#banner-(.*)` dynamic selector case.

**Comment by vmanthos** [comment]
**Related:** https://secure.helpscout.net/conversation/1797479762/328181/

In this case, it's the WP Recipe Maker Premium plugin.

Some of their IDs, e.g. `.wprm-advanced-list-(.*)` are changing dynamically, although this isn't happening on every (uncached_ page load.

**Comment by vmanthos** [comment]
**Related:** https://secure.helpscout.net/conversation/1805729020/329618?folderId=2135277

Flatsome's ` #banner-(.*)` dynamic selector.

**Comment by piotrbak** [comment]
@mehedihasanziku this issue is solved in 3.11. You can use the following filters:
1. This [one](https://github.com/wp-media/wp-rocket/blob/63f649200fa3b5b81ec83812d3448eeeea28e464/inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php#L331) can be used in order not to remove the inline stylesheets based on the style attributes.
2. This [one](https://github.com/wp-media/wp-rocket/blob/63f649200fa3b5b81ec83812d3448eeeea28e464/inc/Engine/Optimization/RUCSS/Controller/UsedCSS.php#L345) can be used in order not to remove inline stylesheets based on the CSS content.
